### ISSUE TITLE
How to use expat for embedded processors

### ISSUE THREAD

**Issue by ataulghalib** [libexpat/libexpat#499]
Hi,
I needed to use expat in an STM32 application written in C. The project is compiled in Embitz in Windows for STM32 target. I tried to follow the example code of e.g. element.c within my porject but start and end element handler never get called. Although if I try to compile it with CMake for windows as separate application and give xml file as parameter to compile *.exe example it parses the xml file correctly.
Do I have to have some special defines for embedded targets as STM32? Initially I was having assertion due Hash salt but then I set the hash salth using XML_SetHashSatl(parser, 1) which helped. But still the start and end element handler never get called when I run my project in embedded target.

**Comment by ataulghalib** [comment]
If you could provide me with some example it would be nice.

**Comment by hartwork** [comment]
I'm happy to help with something concrete, but I don't see a concrete question, problem description, your code, error codes you received, environment description in more detail. PS: Please note that a constant hash salt of 1 will be a security problem if the user can control all or parts of the XML input.

**Comment by ataulghalib** [comment]
Hi,
We have very controlled XMLs so i am not concerned about security I am just
concerned about it working. Following is the part of code I am using.

static void XMLCALL
startElement(void *userData, const XML_Char *name, const XML_Char **atts) {
  int i;
  int *depthPtr = (int *)userData;
  (void)atts;

  for (i = 0; i < *depthPtr; i++)
    putchar('\t');
  printf("%" XML_FMT_STR "\n", name);
  *depthPtr += 1;
}

static void XMLCALL
endElement(void *userData, const XML_Char *name) {
  int *depthPtr = (int *)userData;
  (void)name;

  *depthPtr -= 1;
}

bool readUserProgramFromFlash( void )
{
F_FILE *fp;
char * xmltext = NULL;
XML_Parser parser;

// Open the program xml file for reading
fp = f_open( TEST_PROG_FILE, F_OPEN_MODE_READ );

// If test xml file opened
if( fp != NULL)
{
 char buf[BUFSIZ];
 XML_Parser parser = XML_ParserCreate(NULL);
 int done;
 int depth = 0;

 XML_SetUserData(parser, &depth);
 XML_SetElementHandler(parser, startElement, endElement);
 XML_SetHashSalt(parser, 42);
 do {
   size_t len = fread(buf, 1, sizeof(buf), fp);
   done = len < sizeof(buf);
   if (XML_Parse(parser, buf, (int)len, done) == XML_STATUS_ERROR)
   {
    return false;
   }
 } while (! done);
 XML_ParserFree(parser);
}
return true;
}

So XML file get's opened fine and get's read as well, I have checked it
with step by step debugging in target. But StartElement handler never gets
called.

Br

--
Ataul Ghalib

**Comment by hartwork** [comment]
When I turn your code into a full program (pasted below), `startElement` is called as expected.
My XML file is nothing but `<t></t>`.
Can you share the XML file you are parsing?

```c
#include <expat.h>
#include <stdbool.h>
#include <stdio.h>

#define XML_FMT_STR "s"
#define F_FILE FILE
#define F_OPEN_MODE_READ "r"
#define f_open fopen

static void XMLCALL startElement(void *userData, const XML_Char *name,
                                 const XML_Char **atts) {
  int i;
  int *depthPtr = (int *)userData;
  (void)atts;

  for (i = 0; i < *depthPtr; i++)
    putchar('\t');
  printf("%" XML_FMT_STR "\n", name);
  *depthPtr += 1;
}

static void XMLCALL endElement(void *userData, const XML_Char *name) {
  int *depthPtr = (int *)userData;
  (void)name;

  *depthPtr -= 1;
}

bool readUserProgramFromFlash(const char *TEST_PROG_FILE) {
  F_FILE *fp;
  char *xmltext = NULL;
  XML_Parser parser;

  // Open the program xml file for reading
  fp = f_open(TEST_PROG_FILE, F_OPEN_MODE_READ);

  // If test xml file opened
  if (fp != NULL) {
    char buf[BUFSIZ];
    XML_Parser parser = XML_ParserCreate(NULL);
    int done;
    int depth = 0;

    XML_SetUserData(parser, &depth);
    XML_SetElementHandler(parser, startElement, endElement);
    XML_SetHashSalt(parser, 42);
    do {
      size_t len = fread(buf, 1, sizeof(buf), fp);
      done = len < sizeof(buf);
      if (XML_Parse(parser, buf, (int)len, done) == XML_STATUS_ERROR) {
        return false;
      }
    } while (!done);
    XML_ParserFree(parser);
  }
  return true;
}

int main(int argc, char **argv) {
  int i = 1;
  int exit_code = 0;
  for (; i < argc; i++) {
    printf("Parsing %d of %d: %s...\n", i, argc - 1, argv[i]);
    if (!readUserProgramFromFlash(argv[i])) {
      exit_code = 1;
    }
  }
  return exit_code;
}
```

```console
# gcc -std=c99 -lexpat main.c && ./a.out foo.xml foo.xml 
Parsing 1 of 2: foo.xml...
t
Parsing 2 of 2: foo.xml...
t
```

**Comment by hartwork** [comment]
PS: Please note that with the current logic, `readUserProgramFromFlash` returns `true` even if the file failed to open.

**Comment by ataulghalib** [comment]
Hi,
Yes the logic is not final yet. But as I said I am debugging it so I know
file opening gets fine. The xml file is attached.

I dont know if I have mentioned it already I had to use
#define XML_POOR_ENTROPY
in order to get it compiled. Are you running your code in some embedded
processor? and which one? I am using STM32


And thanks for such a quick response :)
Br
--
Ataul Ghalib

**Comment by ataulghalib** [comment]
And these are the files which I am including in my project

[image: image.png]
--
Ataul Ghalib

**Comment by hartwork** [comment]
> The xml file is attached.
> [..]
> And these are the files which I am including in my project

I don't see any. My guess is that e-mail attachments are ignored.

> I dont know if I have mentioned it already I had to use
> #define XML_POOR_ENTROPY
> in order to get it compiled.

That should only be needed only if you do not have any high quality source of entropy. Without high quality entropy, protection against hash flooding is not possible.

> Are you running your code in some embedded
processor?

No.

**Comment by ataulghalib** [comment]
It does not allow to attach xml files. I try to paste the content

<?xml version="1.0" encoding="UTF-8" standalone="true"?>

-<userprogram>

<model value="10"/>

<name value="ata5"/>

<calibration index="0"/>

<protection value="0"/>


-<fillstep>

<speed value="10"/>

<volume value="125.0"/>

<excess value="0.0"/>

</fillstep>


-<dispensestep>

<speed value="10"/>

<volume value="125.0"/>

<blowout value="100"/>

</dispensestep>


-<notestep>

<text>Tove</text>

</notestep>

</userprogram>

**Comment by ataulghalib** [comment]
It does not allow to attach xml format. I try to paste the content

<?xml version="1.0" encoding="UTF-8" standalone="true"?>
<userprogram>
<model value="10"/>
<name value="ata5"/>
<calibration index="0"/>
<protection value="0"/>
  <fillstep>
  <speed value="10"/>
  <volume value="125.0"/>
  <excess value="0.0"/>
</fillstep>
<dispensestep>
  <speed value="10"/>
  <volume value="125.0"/>
  <blowout value="100"/>
</dispensestep>
<notestep>
  <text>Tove</text>
</notestep>
</userprogram>

**Comment by ataulghalib** [comment]
:( the content also not possible

**Comment by ataulghalib** [comment]
Well I dont think it has anything to do with the xml file. As if I try to compile this code as separate application as you did. It works for me, But if I try to put it in my project code it does not

**Comment by ataulghalib** [comment]
![image](https://user-images.githubusercontent.com/56120259/121898436-0b7e1700-cd2c-11eb-95ec-94b05bc1790b.png)

**Comment by ataulghalib** [comment]
These are the files I am using from expat lib in my project

**Comment by hartwork** [comment]
> It does not allow to attach xml format. I try to paste the content
> Tove

Adding `.txt` to the filename usually helps.

> Well I dont think it has anything to do with the xml file. As if I try to compile this code as separate application as you did. It works for me, But if I try to put it in my project code it does not

That sounds a lot like the problem is outside of expat, no?

**Comment by ataulghalib** [comment]
[ata5.txt](https://github.com/libexpat/libexpat/files/6648627/ata5.txt)

Well the callbacks should still be coming from expat when it hits the start/End of Element. So i think that probably I am missing some setting in expat. Because XML_parse() function takes the whole in, now the code in that file is so big is really hard to follow what should happen.

**Comment by hartwork** [comment]
> [ata5.txt](https://github.com/libexpat/libexpat/files/6648627/ata5.txt)

That file parses successfully with the code I pasted above. So the file is not the problem.

> What I think is that for embedded target expat might some settings some macros or something (I don't know what)

There are no macros or settings like that. STM32 seems to be a 32bit arm architecture. Expat is used on arm (e.g. see https://packages.gentoo.org/packages/dev-libs/expat) and I haven't heard about any arm-specific problems with Expat in recent years. 

Are you using either of the two build systems shipped with Expat to compile Expat?
I'll need a concrete problem to be able to help.

**Comment by ataulghalib** [comment]
No, my project is compiled in Embitz (which uses ARM GCC). I have simply added the following files in the project to be compiled 
![image](https://user-images.githubusercontent.com/56120259/121916744-18efcd00-cd3d-11eb-8298-f9824784010d.png)
I am not building expat separately instead within the project.

**Comment by hartwork** [comment]
>  I am not building expat separately instead within the project.

That makes a build system issue more likely. Can you share your `expat_config.h` file?

**Comment by ataulghalib** [comment]
I am not using any expat_config.h file. All the files I am compiling are in the image I posted above. I could find expat_config.h.cmake file which I think will be extracted to expat_config,h if you use CMake but as mentioned I am not using it.
Should I create expat_config.h myself out expat_config.h.cmake and include in my project?

**Comment by hartwork** [comment]
Unless you patched it away, Expat >=2.4.0 requires an `expat_config.h` file (see #467) to even compile. Expat has two supported build systems. If you go off supported terrain, you need to study at least one of the two build system to see how to properly configure Expat. It would be no surprise if your Expat is not behaving as expected. My advice would be stay as close as possible to one of these two build systems. You can derive `expat_config.h` from either `expat_config.h.cmake` or from `expat_config.h.in` manually, they are similar.

Three more things I'd like to ask:
- Which version of Expat are you using?
- Are you 100% sure that the parse returns success? The code above had no reporting about that.
- Are all Expat files 100% vanilla or have you patched anything?

**Comment by ataulghalib** [comment]
#define XML_MAJOR_VERSION 2
#define XML_MINOR_VERSION 3
#define XML_MICRO_VERSION 0

**Comment by ataulghalib** [comment]
Okay now I am using 2.4.1. I think it is related to expa_config.h file. Attached is the config file which I am using now. XML_parse() does not return success it returns error XML_ERROR_NO_ELEMENTS whereas my xml file does have elements
[expat_config.txt](https://github.com/libexpat/libexpat/files/6650069/expat_config.txt)

**Comment by hartwork** [comment]
> Okay now I am using 2.4.1.

Thank you!

> I think it is related to expa_config.h file. Attached is the config file which I am using now. XML_parse() does not return success it returns error XML_ERROR_NO_ELEMENTS whereas my xml file does have elements

Well, it's no wonder that the start handler is not called for no elements, right? :smiley: 
Could it be that you are pushing the empty string rather than actual data to the parser? Maybe `fread` always returns 0?

> [expat_config.txt](https://github.com/libexpat/libexpat/files/6650069/expat_config.txt)

Thanks!
What catches my eye here are:
- `XML_ATTR_INFO` — try inverting that one please, just to see if things change. Probably not any issue.
- Try without `off_t OFF_T` and `size_t SIZE_T` please, just to see if things change. Probably not any issue.

**Comment by ataulghalib** [comment]
okay finally now I got it working to the level that it is calling the start/end element handlers now. Thanks for your support. I must you have been very patient and helping. Thanks a lot. I will get back to you if I face any other hurdle.
Chao

**Comment by hartwork** [comment]
Glad to hear. I will close the ticket as "done" then.
